include/deal.II/base/big_mpi_compat.h

```diff
@@ -0,0 +1,338 @@
+
+      MPI_Count size_old;
+      ierr = MPI_Type_size_x(oldtype, &size_old);
+
+      MPI_Count n_chunks          = count / max_signed_int;
+      MPI_Count n_bytes_remainder = count % max_signed_int;
+
+      MPI_Datatype chunks;
+      ierr = MPI_Type_vector(
+        n_chunks, max_signed_int, max_signed_int, oldtype, &chunks);
+      if (ierr != MPI_SUCCESS)
+        {
+          std::cerr << "MPI_Type_vector() call failed!" << std::endl;
+          return ierr;
+        }
+
+      MPI_Datatype remainder;
+      ierr = MPI_Type_contiguous(n_bytes_remainder, oldtype, &remainder);
+      if (ierr != MPI_SUCCESS)
+        {
+          std::cerr << "MPI_Type_contiguous() call failed!" << std::endl;
+          return ierr;
+        }
+
+      int          blocklengths[2]  = {1, 1};
+      MPI_Aint     displacements[2] = {0,
+                                   static_cast<MPI_Aint>(n_chunks) * size_old *
+                                     max_signed_int};
+      MPI_Datatype types[2]         = {chunks, remainder};
+      ierr =
+        MPI_Type_create_struct(2, blocklengths, displacements, types, newtype);
+      if (ierr != MPI_SUCCESS)
+        {
+          std::cerr << "MPI_Type_contiguous() call failed!" << std::endl;
+          return ierr;
+        }
+      ierr = MPI_Type_commit(newtype);
+      if (ierr != MPI_SUCCESS)
+        {
+          std::cerr << "MPI_Type_contiguous() call failed!" << std::endl;
+          return ierr;
+        }
+
+      ierr = MPI_Type_free(&chunks);
+      if (ierr != MPI_SUCCESS)
+        {
+          std::cerr << "MPI_Type_contiguous() call failed!" << std::endl;
+          return ierr;
+        }
+      ierr = MPI_Type_free(&remainder);
+      if (ierr != MPI_SUCCESS)
+        {
+          std::cerr << "MPI_Type_contiguous() call failed!" << std::endl;
+          return ierr;
+        }
+
+#    ifndef MPI_COMPAT_SKIP_SIZE_CHECK
+      MPI_Count size_new;
+      ierr = MPI_Type_size_x(*newtype, &size_new);
+
+      if (size_old * count != size_new)
+        {
+          std::cerr
+            << "MPI_Type_contiguous_c() produced an invalid result. Expected = "
+            << size_old << " * " << count << " = " << size_old * count
+            << " but received " << size_new << std::endl;
+          return MPI_ERR_INTERN;
+        }
+#    endif
+
+      return MPI_SUCCESS;
+    }
+}
+
+/**
+ * Send a package to rank @p dest with a (possibly large) @p count.
+ *
+ * See the MPI 4.x standard for details.
+ */
+inline int
+MPI_Send_c(const void * buf,
+           MPI_Count    count,
+           MPI_Datatype datatype,
+           int          dest,
+           int          tag,
+           MPI_Comm     comm)
+{
+  if (count <= BigMPICompat::mpi_max_count)
+    return MPI_Send(buf, count, datatype, dest, tag, comm);
+
+  MPI_Datatype bigtype;
+  int          ierr;
+  ierr = MPI_Type_contiguous_c(count, datatype, &bigtype);
+  if (ierr != MPI_SUCCESS)
+    return ierr;
+  ierr = MPI_Type_commit(&bigtype);
```

Comment:
Do you know how that can be expensive as this is not a collective call? Or do I also need to use it in a collective call?
I can try this on Frontera on 100k ranks but I have only 1 or 2 MPI implementations I can use.

Any suggestions?

Reply:
I was not trying to suggest that this particular case is expensive (most likely it is not), I just recall that at some point I did see a similar function in my profiler (but there was some synchronization issue). I believe that no synchronization is needed here (as it would defeat the whole idea of a single send operation), so I would just go ahead. Maybe you can have an eye on this in case it pops up as a "significant slowdown" in the near future.

Reply:
I will run a large test in a couple days to see if it works correctly.

Reply:
It's a purely local call. It's hard to imagine that being expensive unless you are creating millions of these structures...